development

harsh-jagtap-josh opened the pull request.

harsh-jagtap-josh committed:
add changes

harsh-jagtap-josh committed:
trial commit

harsh-jagtap-josh committed:
add changes for file structuring

harsh-jagtap-josh committed:
add minor changes

RushikeshMarkad16 commented on internal/repository/init.go:
```diff
@@ -0,0 +1,36 @@
+		return nil, err
+	}
+
+	// log.Println("Database connected!")
```
Remove unnecessary logs

harsh-jagtap-josh replied:
okay removed all unnecessary prints and commented logs throughout the code..

RushikeshMarkad16 commented on internal/repository/init.go:
```diff
@@ -0,0 +1,36 @@
+
+func ConnectDB() (*sql.DB, error) {
+	const (
+		host     = "localhost"
```
create separate struct for db config. function as well

harsh-jagtap-josh replied:
created a struct for DB data in same file, and renamed the function as well..

RushikeshMarkad16 commented on internal/repository/init.go:
```diff
@@ -0,0 +1,36 @@
+func ConnectDB() (*sql.DB, error) {
+	const (
+		host     = "localhost"
+		port     = 5432
```
use env variables instead. do not hardcode

harsh-jagtap-josh replied:
okay.. created an env file and appended all the changes..

RushikeshMarkad16 commented on internal/repository/init.go:
```diff
@@ -0,0 +1,36 @@
+	)
+
+	db, err := sql.Open("postgres", psqlconn)
+	if err != nil {
```
logging is must whenever err is encountered. Do this change everywhere

harsh-jagtap-josh replied:
I'll work on the logging functionality soon..

harsh-jagtap-josh replied:
added logs.

RushikeshMarkad16 commented on internal/repository/postgres/base.go:
```diff
@@ -0,0 +1,7 @@
+package postgres
```
you can rename this package as repo.

harsh-jagtap-josh replied:
okay, renamed the package as asked

RushikeshMarkad16 commented on internal/repository/workers.go:
file name should be worker.go

harsh-jagtap-josh replied:
renamed the file..

RushikeshMarkad16 commented on internal/repository/repo.go:
```diff
@@ -0,0 +1,17 @@
+// RepositoryTransaction interface holds transaction specific methods
+type RepositoryTransaction interface {
+	// return a transaction from a gorm connection
+	BeginTx(ctx context.Context) (Transaction, error)
```
Do we need transactions at this stage?

harsh-jagtap-josh replied:
no we don't.. thats why it was already set to nil in the worker handler where it was called.. still removed it

RushikeshMarkad16 commented on internal/repository/postgres/worker.go:
```diff
@@ -0,0 +1,30 @@
+	}
+}
+
+func (ps *workerStore) GetWorkerByID(ctx context.Context, tx repository.Transaction, workerID int64) (repository.Worker, error) {
```
what is ps in receiver?

harsh-jagtap-josh replied:
sorry, I meant ws - worker service, had `ps` due to replicating Sagar's code.

RushikeshMarkad16 commented on internal/repository/postgres/worker.go:
```diff
@@ -0,0 +1,30 @@
+}
+
+func (ps *workerStore) GetWorkerByID(ctx context.Context, tx repository.Transaction, workerID int64) (repository.Worker, error) {
+	query := `SELECT * FROM Workers WHERE id = $1;`
```
use constant

harsh-jagtap-josh replied:
okay, haven't made the change, will make the changes when there will be multiple apis

RushikeshMarkad16 replied:
why cant we make change for single API?

harsh-jagtap-josh replied:
made the  changes.

RushikeshMarkad16 commented on internal/repository/postgres/worker.go:
```diff
@@ -0,0 +1,30 @@
+	row := ps.DB.QueryRow(query, workerID)
+
+	var worker repository.Worker
+	err := row.Scan(&worker.ID, &worker.Name, &worker.ContactNumber, &worker.Email, &worker.Gender,
```
do we have any other way to do this?

harsh-jagtap-josh replied:
mostly this way is used, i'll try to find for other ways..

RushikeshMarkad16 replied:
Try to map directly from db to struct without passing each field explicitly

harsh-jagtap-josh replied:
there is no function to do this in the standard `sql` package, might have to use external packages like `sqlx` here..

RushikeshMarkad16 replied:
RIght. You can go ahead with it.

harsh-jagtap-josh replied:
okay..

RushikeshMarkad16 commented on internal/repository/workers.go:
```diff
@@ -0,0 +1,36 @@
+	Unknown Gender = "Unknown"
+)
+
+type Worker struct {
```
move structs to domain.go

harsh-jagtap-josh replied:
okay, made the changes.. moved all structs to dto folder in `internal/pkg` folder... dto is like data transfer objects

RushikeshMarkad16 replied:
Research more about what is significance of internal/pkg

harsh-jagtap-josh replied:
yes.. done that

RushikeshMarkad16 commented on internal/pkg/middleware.go:
```diff
@@ -0,0 +1,57 @@
+	"net/http"
+)
+
+type httpErrorResponse struct {
```
name can be ErrorResponse

harsh-jagtap-josh replied:
okay renamed this and other response structs..

RushikeshMarkad16 commented on cmd/main.go:
```diff
@@ -0,0 +1,34 @@
+	router := api.NewRouter(services)
+
+	srv := &http.Server{
+		Addr:    fmt.Sprintf(":%d", 8080),
```
use env var for port number

harsh-jagtap-josh replied:
yes made the changes..

RushikeshMarkad16 commented on cmd/main.go:
```diff
@@ -0,0 +1,34 @@
+		Handler: router,
+	}
+	fmt.Println("Server up and running on port 8080")
+	log.Fatal(srv.ListenAndServe())
```
catch error from ListenAndServe

harsh-jagtap-josh replied:
okay, i'll add error handling here

RushikeshMarkad16 commented on internal/api/router.go:
```diff
@@ -0,0 +1,15 @@
+func NewRouter(deps app.Dependencies) *mux.Router {
+	router := mux.NewRouter()
+
+	WorkersRouter := router.PathPrefix("/workers").Subrouter()
```
why WorkersRouter is exported?

harsh-jagtap-josh replied:
there was no need of exporting here, sorry didn't notice that.. made the changes

RushikeshMarkad16 commented on internal/api/router.go:
```diff
@@ -0,0 +1,15 @@
+	router := mux.NewRouter()
+
+	WorkersRouter := router.PathPrefix("/workers").Subrouter()
+	WorkersRouter.HandleFunc("/{worker_id}", HandleWorkerByID(deps.WorkerService)).Methods("GET")
```
use http.MethidGet

harsh-jagtap-josh replied:
okay, will use it now on instead of hard-coding

RushikeshMarkad16 commented on internal/api/worker.go:
```diff
@@ -0,0 +1,36 @@
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		vars := mux.Vars(r)
+		rawProductID := vars["worker_id"]
```
what does rawProductID mean?

harsh-jagtap-josh replied:
sorry the 'raw' here basically means the workerID retrieved from the query slugs, which we are later going to convert to integer, and these integer variables will be named workerID.. had productID due to trying to replicate Sagar's code..

RushikeshMarkad16 commented on internal/api/worker.go:
```diff
@@ -0,0 +1,36 @@
+		ctx := r.Context()
+		vars := mux.Vars(r)
+		rawProductID := vars["worker_id"]
+		productID, err := strconv.Atoi(rawProductID)
```
what is productID??

harsh-jagtap-josh replied:
Sorry, I meant worker ID.. stated the reason for this in prior comments..

RushikeshMarkad16 commented on internal/api/worker.go:
```diff
@@ -0,0 +1,36 @@
+
+		response, err := workerSvc.GetWorkerByID(ctx, nil, int64(productID))
+		if err != nil {
+			middleware.HandleErrorResponse(w, "error occured while fetching from database", http.StatusInternalServerError)
```
give descriptive error

harsh-jagtap-josh replied:
already added a description, how should I describe it more, or change the existing error?

RushikeshMarkad16 replied:
Error should be like:
error occurred while fetching record for worker with id: <worker_id>

harsh-jagtap-josh replied:
got this, I changed it as well in the next commit i made..

RushikeshMarkad16 commented on internal/api/worker.go:
```diff
@@ -0,0 +1,36 @@
+		}
+
+		if response.ID == 0 {
+			middleware.HandleErrorResponse(w, "error occured: no worker found with ID", http.StatusInternalServerError)
```
mention id as well in error

harsh-jagtap-josh replied:
okay.. added it in the error string provided

RushikeshMarkad16 replied:
By referencing id i meant value for the key id .

harsh-jagtap-josh replied:
i added ID in error in the commit..

RushikeshMarkad16 commented on internal/pkg/middleware.go:
```diff
@@ -0,0 +1,57 @@
+	ErrorMessage string `json:"error_message"`
+}
+
+type httpSuccessResponse struct {
```
same here

harsh-jagtap-josh replied:
made the changes

RushikeshMarkad16 commented on internal/app/worker/service.go:
```diff
@@ -0,0 +1,34 @@
+func (ps *service) GetWorkerByID(ctx context.Context, tx repository.Transaction, productID int64) (repository.Worker, error) {
+	productInfoDB, err := ps.workerRepo.GetWorkerByID(ctx, tx, productID)
+	if err != nil {
+		return repository.Worker{}, nil
```
Why are we returning nil if there is error?

harsh-jagtap-josh replied:
sorry didn't notice that, made it to return an empty worker obj and the error which will be further handled in the handler..

RushikeshMarkad16 commented on internal/app/worker/service.go:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	if productInfoDB.ID == 0 {
+		return repository.Worker{}, nil
```
Why are we returning nil if there is error?

harsh-jagtap-josh replied:
there is no error here, just the result is empty.. like there is no worker with given ID.. this error is handled in api/worker.go

RushikeshMarkad16 replied:
Isnt this an error when we are trying to fetch very specific worker with given id?

harsh-jagtap-josh replied:
yes it is technically an error, but the error is handled already with a different message just above this.. so this will never be hit, so I might have to add specific error in repo file like sql.ErrNoRows which specifically signifies that the query was executed successfully but there are no rows in the output.. and handle it in service accordingly..

RushikeshMarkad16 commented on internal/api/worker.go:
why this file is a part of API and not app?

harsh-jagtap-josh replied:
because it only contains handler function, and service part is executed in app/worker/service.go

RushikeshMarkad16 replied:
handlers are also specific to the entity right? Its good idea to keep it in same folder

harsh-jagtap-josh replied:
I have `routes.go` file and handlers for all entities in seperate files, seperate file for each entity containing only handlers for that entity.. rest logic is in services and repo..

harsh-jagtap-josh committed:
add minor changes

harsh-jagtap-josh committed:
add .env functionality, error handling, renaming files based on convention, moving structs to dto folder, removing transaction objs used in repo and postgres folders

harsh-jagtap-josh committed:
fix: removed unused modules from go.mod file

RushikeshMarkad16 commented on .env:
avoid pushing your actual env file to github. You can add it to .gitignore.
Instead add example file here

harsh-jagtap-josh replied:
Okay will do this

RushikeshMarkad16 commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+func main() {
+	// ctx := context.Background()
+
+	godotenv.Load()
```
Please provide attention to details. Load function returns error. Why didn't you watched it here?

harsh-jagtap-josh replied:
i'll add it here.. didn't notice

harsh-jagtap-josh replied:
made the changes..

RushikeshMarkad16 commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+	godotenv.Load()
+	sqlDB, err := repository.InitDB()
+	if err != nil {
+		fmt.Println("error connecting to database" + err.Error())
```
use logger everywhere

harsh-jagtap-josh replied:
done.

RushikeshMarkad16 commented on go.mod:
```diff
@@ -0,0 +1,10 @@
+go 1.23.4
+
+require (
+	github.com/caarlos0/env/v11 v11.3.1
```
Are we using this anywhere?

harsh-jagtap-josh replied:
no we are not.. even removed this in the last commit..

RushikeshMarkad16 commented on internal/app/worker/domain.go:
```diff
@@ -0,0 +1,29 @@
+	Password        string    `db:"password"`
+	Sectors         string    `db:"sectors"`
+	Skills          string    `db:"skills"`
+	Location        *int      `db:"location,omitempty"`
```
Why did we need to use pointer here?

RushikeshMarkad16 replied:
same for other fields

RushikeshMarkad16 commented on internal/repository/worker.go:
```diff
@@ -0,0 +1,11 @@
+)
+
+type WorkerStorer interface {
+	GetWorkerByID(ctx context.Context, productID int64) (dto.Worker, error)
```
what productID signifies here?

harsh-jagtap-josh replied:
updated the code..

RushikeshMarkad16 commented on internal/repository/postgres/worker.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func (ws *workerStore) GetWorkerByID(ctx context.Context, workerID int64) (dto.Worker, error) {
+	query := `SELECT * FROM Workers WHERE id = $1;`
```
is table name Workers with capital W?

harsh-jagtap-josh replied:
yes it Capital in table name

RushikeshMarkad16 replied:
Is it a good naming convention for db tables?

harsh-jagtap-josh replied:
no it is not a good convention.. made the changes.

RushikeshMarkad16 commented on internal/pkg/middleware.go:
```diff
@@ -0,0 +1,57 @@
+
+func HandleErrorResponse(w http.ResponseWriter, errMessage string, errStatusCode int) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(errStatusCode)
```
check whether this status code persists always or it changes according to conditions given below?

harsh-jagtap-josh replied:
it's not static, it uses whatever status code is provided in it when its called..

RushikeshMarkad16 replied:
Are u sure? Have u tested it?

harsh-jagtap-josh replied:
yes i have tested it..

harsh-jagtap-josh committed:
add major changes
- create zap logger, log all errors
- create new apperrors folder, holding all error messages and other functions
- re-strucuture project
- remove api folder and move all files to app folder
- add changes in handlers, updating http error messages, api responses, error messages and logging
- remove .env from git
- remove db init from repo folder, add to config in base directory

RushikeshMarkad16 commented on .vscode/settings.json:
Is this file relevant to MR?

harsh-jagtap-josh replied:
No, removed the file..

RushikeshMarkad16 commented on cmd/main.go:
```diff
@@ -0,0 +1,48 @@
+
+	sqlDB, err := db.InitDB()
+	if err != nil {
+		logger.Errorw(ctx, "failed to start a database connection", zap.Error(err))
```
you can use word establish or create instead of start

harsh-jagtap-josh replied:
made the changes..

RushikeshMarkad16 commented on config.go:
```diff
@@ -0,0 +1,53 @@
+	"github.com/joho/godotenv"
+)
+
+type DbData struct {
```
Can we use DbConfig name here?

harsh-jagtap-josh replied:
done.

RushikeshMarkad16 commented on config.go:
```diff
@@ -0,0 +1,53 @@
+	)
+
+	db, err := sql.Open("postgres", psqlconn)
+	if err != nil {
```
add log here

harsh-jagtap-josh replied:
okay, must have forgot in this file.

RushikeshMarkad16 commented on config.go:
```diff
@@ -0,0 +1,53 @@
+	// Check if the database connection is working
+	err = db.Ping()
+	if err != nil {
+		db.Close()
```
add log whenever err is encountered. Do this change everywhere.

harsh-jagtap-josh replied:
made the changes..

RushikeshMarkad16 commented on internal/app/router.go:
```diff
@@ -0,0 +1,17 @@
+func NewRouter(deps Dependencies) *mux.Router {
+	router := mux.NewRouter()
+
+	workersRouter := router.PathPrefix("/workers").Subrouter()
```
can we use worker instead of workers

harsh-jagtap-josh replied:
ok made the changes..

RushikeshMarkad16 commented on internal/app/worker/handler.go:
```diff
@@ -0,0 +1,48 @@
+		id := vars["worker_id"]
+		workerID, err := strconv.Atoi(id)
+		if err != nil {
+			logger.Errorw(ctx, apperrors.ErrConvertIdToInt, zap.Error(err), zap.String("ID", id))
```
this should be ErrInvalidWorkerId

harsh-jagtap-josh replied:
okay, made the changes..

RushikeshMarkad16 commented on internal/app/worker/handler.go:
```diff
@@ -0,0 +1,48 @@
+		}
+
+		response, err := workerSvc.GetWorkerByID(ctx, workerID)
+		if errors.Is(err, sql.ErrNoRows) {
```
you should check this err at db level not here

harsh-jagtap-josh replied:
it was already handled in db level, but again we return the same err obj, and have to identify which kind of error it was in handler again, so removed it from db level..

RushikeshMarkad16 commented on internal/app/worker/handler.go:
```diff
@@ -0,0 +1,48 @@
+		}
+
+		response, err := workerSvc.GetWorkerByID(ctx, workerID)
+		if errors.Is(err, sql.ErrNoRows) {
```
Shoudn't we check if err is present or not before comparison

harsh-jagtap-josh replied:
even if we check it won't affect the current working, will be the same as now.. we would just be adding an additional block of code which might have no effect..

RushikeshMarkad16 replied:
But you are comparing error for every call even if err is nil. Is this the right way?

harsh-jagtap-josh replied:
yes, this isn't right I realized it later, made the changes here..

RushikeshMarkad16 commented on internal/pkg/apperrors/errors.go:
```diff
@@ -0,0 +1,13 @@
+
+var (
+	ErrInternalServerError = errors.New("internal server error")
+	ErrInvalidRequestParam = errors.New("invalid request param")
```
Isn't this very generalized error msg?

harsh-jagtap-josh replied:
yes, but will be used almost every api, so thought to keep it in seperate file.. should i remove this?

RushikeshMarkad16 replied:
You should mention which param is invalid

harsh-jagtap-josh replied:
I think this will be mentioned in the err object that we have passed while logging this error, so it will be covered there..

RushikeshMarkad16 commented on internal/pkg/apperrors/http_response.go:
Why do we require this file?

harsh-jagtap-josh replied:
had a piece of code here, but moved it to middleware file, so now this file is not required.. removed the file..

RushikeshMarkad16 commented on logfile:
Do we need this file here?

harsh-jagtap-josh replied:
No, it is auto generated by uber/zap ..

RushikeshMarkad16 replied:
But be careful while addding it to git. Try to avoid using `git add .`

RushikeshMarkad16 commented on internal/repo/worker.go:
```diff
@@ -0,0 +1,41 @@
+import (
+	"context"
+	"database/sql"
+	// "github.com/harsh-jagtap-josh/RozgarLink/internal/repo"
```
remove unwanted code

harsh-jagtap-josh replied:
done..

RushikeshMarkad16 commented on internal/repo/worker.go:
```diff
@@ -0,0 +1,41 @@
+	// repo "github.com/harsh-jagtap-josh/RozgarLink/internal/repository"
+)
+
+const SelectWorkerByIDQuery = "SELECT id, name, contact_number, email, gender, sectors, skills, location, is_available, rating, total_jobs_worked, created_at, updated_at, language FROM workers WHERE id = $1;"
```
Why is this exported? Are we using this outside package?

harsh-jagtap-josh replied:
no, we are not using it for now..

RushikeshMarkad16 commented on internal/repo/worker.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+const SelectWorkerByIDQuery = "SELECT id, name, contact_number, email, gender, sectors, skills, location, is_available, rating, total_jobs_worked, created_at, updated_at, language FROM workers WHERE id = $1;"
+const UpdateWorkerByIDQuery = ""
```
Is this related to this PR of GetWorkerById?

harsh-jagtap-josh replied:
no but will be adding this function, and might need it then, so created it..

RushikeshMarkad16 replied:
but PR should contain only related changes

RushikeshMarkad16 commented on internal/pkg/apperrors/worker.go:
We can keep all errors in one file, no need of separate files for each entity

harsh-jagtap-josh replied:
okay, there will be seperate error messages for each of the entity thats why thought of keeping all them in seperate files.. should i move them to single file?

RushikeshMarkad16 replied:
As of now, yes

harsh-jagtap-josh replied:
okay, made the changes..

RushikeshMarkad16 commented on internal/pkg/apperrors/worker.go:
```diff
@@ -0,0 +1,16 @@
+const ErrFetchFromDb = "error occurred while fetching from database"
+const ErrEmailAlready = "email already exists"
+const ErrFailedToFetchWorker = "failed to fetch worker"
+const ErrFailedToCreateWorker = "failed to create user"
```
be consistent in error messages. here you are using user and worker for other

harsh-jagtap-josh replied:
made the changes..

harsh-jagtap-josh committed:
add minor changes, add log messages, updating error messages, removing comments, renaming structs and functions

harsh-jagtap-josh committed:
add changes
- move all errors to single file
- update error handling in worker handler
- remove logfile from git

RushikeshMarkad16 commented on go.mod:
```diff
@@ -0,0 +1,14 @@
+
+require (
+	go.uber.org/multierr v1.11.0 // indirect
+	go.uber.org/zap v1.27.0 // indirect
```
You are using zap directly. How is its indirect here?
Have habit of performing `go mod tidy` frequently to update dependencies.

harsh-jagtap-josh replied:
okay

RushikeshMarkad16 commented on internal/app/worker/service.go:
```diff
@@ -0,0 +1,30 @@
+	}
+}
+
+func (ws *service) GetWorkerByID(ctx context.Context, productID int) (repo.Worker, error) {
```
Please update everywhere if there are any review comments. Ig this is the 3 or 4th time I am asking to remove productID from code.

harsh-jagtap-josh replied:
yes. made the changes

RushikeshMarkad16 commented on internal/pkg/apperrors/errors.go:
```diff
@@ -0,0 +1,32 @@
+
+// Worker Errors
+
+const ErrNoWorkerExists = "no worker found with id"
```
errors should be of type error right?

harsh-jagtap-josh replied:
these are actually just error messages that we are using for logging, the actual error is directly logged..

harsh-jagtap-josh committed:
fix: mod file, worker service file

RushikeshMarkad16 commented on internal/app/worker/handler.go:
```diff
@@ -0,0 +1,47 @@
+
+		response, err := workerSvc.GetWorkerByID(ctx, workerID)
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
```
Move this to db

harsh-jagtap-josh committed:
add changes
- remove all dto files
- use sqlx instead of sql in repo functions
- use service layer structs in service returns
- use mapper for converting structs
- move error handling from service to repo
- add db tags to repo domain
- rename errors variables from Err to Msg because they are error messages and not errors
- update error handling in handler

harsh-jagtap-josh committed:
fix: remove unnecessary comments

harsh-jagtap-josh committed:
minor change: change status code & error message for fetch api

harsh-jagtap-josh committed:
fix: renaming service functions for consistency
